backend/db.js

```diff
@@ -0,0 +1,106 @@
+import { DataSource } from 'apollo-datasource'
+import crypto from 'crypto'
+
+export class Post {
+    constructor(data) {
+        this.id = crypto.randomBytes(16).toString('hex')
+        Object.assign(this, data)
+    }
+}
+
+export class User {
+    constructor(data) {
+        this.id = crypto.randomBytes(16).toString('hex')
+        Object.assign(this, data)
+    }
+}
+
+export class InMemoryDataSource extends DataSource {
+    constructor() {
+        super()
+        this.posts = []
+        this.users = []
+        this.initDummyUsers()
+        this.initDummyPosts()
+    }
+
+    initialize(...args) {
+        //Warum wird diese Methode so oft ausgeführt?
+    }
+
+    createPost(data) {
+        const author = this.users.find(user => user.name === data.post.author.name)
+        const newPost = new Post({
+            title: data.post.title,
+            votes: [],
+            author: author
+        })
+        this.posts.push(newPost)
+        const newUsers = this.users.map(user => user.id === newPost.author.id ? this.addPostToAuthor(newPost, user) : user)
+        this.users = [...newUsers]
+
+        return newPost
+    }
+
+    upvotePost(data) {
+        const voter = this.users.find(user => user.name === data.voter.name)
+        const newPosts = this.posts.map(post => post.id === data.id && post.votes.find(user => user.name === voter.name) === undefined ? this.addVoter(post, voter) : post)
```

Comment:
This line is too long to really understand what's going on. I suggest the following solution:

1. Implement a generic `updatePost` method.
2. When you update votes in your resolvers, you find a post first, then update it's votes and then call `updatePost(updatedPost)`